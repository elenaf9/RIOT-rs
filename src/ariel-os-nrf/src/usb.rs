use ariel_os_debug::log::debug;
use embassy_nrf::{
    bind_interrupts, peripherals,
    usb::{
        self,
        vbus_detect::{self, HardwareVbusDetect},
        Driver,
    },
};

#[cfg(context = "nrf52")]
bind_interrupts!(struct Irqs {
    USBD => usb::InterruptHandler<peripherals::USBD>;
    POWER_CLOCK => vbus_detect::InterruptHandler;
});

#[cfg(context = "nrf5340")]
bind_interrupts!(struct Irqs {
    USBD => usb::InterruptHandler<peripherals::USBD>;
    USBREGULATOR => vbus_detect::InterruptHandler;
});

pub type UsbDriver = Driver<'static, peripherals::USBD, HardwareVbusDetect>;

pub fn init() {
    // nrf52840
    let clock: embassy_nrf::pac::CLOCK = unsafe { core::mem::transmute(()) };

    debug!("nrf: enabling ext hfosc...");
    clock.tasks_hfclkstart.write(|w| unsafe { w.bits(1) });
    while clock.events_hfclkstarted.read().bits() != 1 {}
}

pub fn driver(peripherals: &mut crate::OptionalPeripherals) -> UsbDriver {
    let usbd = peripherals.USBD.take().unwrap();
    Driver::new(usbd, Irqs, HardwareVbusDetect::new(Irqs))
}
