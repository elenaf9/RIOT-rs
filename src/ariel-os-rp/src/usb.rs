use embassy_rp::{
    bind_interrupts, peripherals,
    usb::{Driver, InterruptHandler},
};

bind_interrupts!(struct Irqs {
    USBCTRL_IRQ => InterruptHandler<peripherals::USB>;
});

pub type UsbDriver = Driver<'static, peripherals::USB>;

pub fn driver(peripherals: &mut crate::OptionalPeripherals) -> UsbDriver {
    let usb = peripherals.USB.take().unwrap();
    Driver::new(usb, Irqs)
}
