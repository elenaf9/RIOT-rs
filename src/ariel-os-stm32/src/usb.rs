use embassy_stm32::{bind_interrupts, peripherals, usb, usb::Driver};

bind_interrupts!(struct Irqs {
    USB_LP => usb::InterruptHandler<peripherals::USB>;
});

pub type UsbDriver = Driver<'static, peripherals::USB>;

pub fn driver(peripherals: &mut crate::OptionalPeripherals) -> UsbDriver {
    let usb = peripherals.USB.take().unwrap();
    let dp = peripherals.PA12.take().unwrap();
    let dm = peripherals.PA11.take().unwrap();

    Driver::new(usb, Irqs, dp, dm)
}
