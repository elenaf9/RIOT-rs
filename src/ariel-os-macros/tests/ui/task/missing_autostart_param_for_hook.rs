#![no_main]
#![feature(impl_trait_in_assoc_type)]
#![feature(used_with_arg)]

// As the macro will fail, this import will not get used
#[allow(unused_imports)]
use ariel_os::usb::UsbBuilderHook;

// FAIL: using hooks require the task to be autostart
#[ariel_os::task(usb_builder_hook)]
async fn main() {}
