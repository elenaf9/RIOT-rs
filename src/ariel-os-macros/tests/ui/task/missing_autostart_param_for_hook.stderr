error: custom attribute panicked
  --> tests/ui/task/missing_autostart_param_for_hook.rs:10:1
   |
10 | #[ariel_os::task(usb_builder_hook)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: the task must be `autostart` to instantiate hooks
